File: api/lib/triage.js
// US-3: тревожные признаки и категория срочности. Чистые правила, без модели,
// чтобы результат был воспроизводим и в DEMO_MODE, и с живым ключом.

const NEGATIVE = /^(нет|нету|не|no|none|-|—|нет\.|ничего|отрицаю)$/i;

const RED_FLAGS = [
  { re: /(бол|дав|жж|жм|сдав).{0,20}(груд|за\s*груд)|груд.{0,15}(бол|дав|жж)/i, label: 'боль или давление в груди' },
  { re: /(одышк|задыха|не могу дышать|нехватка воздуха|тяжело дышать)/i, label: 'одышка' },
  { re: /(слабость в (рук|ног)|онемел|отнял[оа]сь|парал)/i, label: 'слабость или онемение конечности' },
  { re: /(наруш.{0,10}реч|невнятн.{0,10}реч|не могу говорить|перекос.{0,10}лиц)/i, label: 'нарушение речи или асимметрия лица' },
  { re: /(обморок|потер[ья].{0,10}сознан|теря[юл].{0,10}сознан)/i, label: 'потеря сознания' },
  { re: /(кровотеч|рвота с кровью|кровь в стуле|сильно кровит)/i, label: 'кровотечение' },
  { re: /(судорог|припадок)/i, label: 'судороги' },
  { re: /(темпер|t|т)\D{0,10}(39|40|41)/i, label: 'температура 39 и выше' },
  { re: /(суицид|покончить с собой|не хочу жить)/i, label: 'суицидальные мысли' },
];

export function parsePain(raw) {
  const m = String(raw || '').match(/\d{1,2}/);
  if (!m) return null;
  const n = Number(m[0]);
  if (Number.isNaN(n)) return null;
  return Math.max(0, Math.min(10, n));
}

export function detectRedFlags(answers) {
  const haystack = Object.entries(answers)
    .filter(([key, val]) => !(key === 'redflags' && NEGATIVE.test(String(val).trim())))
    .map(([, val]) => String(val))
    .join(' \n ');

  const found = [];
  for (const flag of RED_FLAGS) {
    if (flag.re.test(haystack)) found.push(flag.label);
  }
  return found;
}

// Lookbehind, а не : в JS границы слова не работают с кириллицей, и без
// проверки предыдущей буквы «понедельник» попадает в «недел», а «погода» — в «год».
const LONG_DURATION = /(?<![а-яё])(недел|месяц|год|давно|постоянн|хроническ)/i;

export const URGENCY = {
  emergency: { code: 'emergency', label: 'Неотложно', rank: 0 },
  today: { code: 'today', label: 'Сегодня', rank: 1 },
  planned: { code: 'planned', label: 'Плановый приём', rank: 2 },
  selfcare: { code: 'selfcare', label: 'Самопомощь', rank: 3 },
};

export function classify(answers) {
  const redFlags = detectRedFlags(answers);
  const pain = parsePain(answers.pain);
  const duration = String(answers.duration || '');
  const all = Object.values(answers).join(' ');

  if (redFlags.length) {
    return { urgency: URGENCY.emergency, redFlags, reason: 'Найдены тревожные признаки: ' + redFlags.join(', ') };
  }
  if ((pain !== null && pain >= 7) || /(38[.,]?\d?|темпер.{0,15}38)/i.test(all) || /внезапн|резко начал/i.test(duration)) {
    return { urgency: URGENCY.today, redFlags, reason: 'Выраженная боль или острое начало — осмотр в течение суток.' };
  }
  if ((pain !== null && pain >= 3) || LONG_DURATION.test(duration)) {
    return { urgency: URGENCY.planned, redFlags, reason: 'Умеренные или длительные симптомы — плановый приём.' };
  }
  return { urgency: URGENCY.selfcare, redFlags, reason: 'Тревожных признаков нет, интенсивность низкая — наблюдение и самопомощь.' };
}

File: rules/triage.py
"""Классификация срочности. Детерминированные правила, без модели.

Всё, что оценивается на отборе, считается здесь: тогда демо-режим и живой
ключ дают одинаковый результат, и проверяющий видит ровно то же, что мы.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, asdict
from datetime import date

import duration as dur

NEGATIVE = re.compile(r"^(нет|нету|не|no|none|-|—|ничего|отрицаю|нет\.)$", re.I)

RED_FLAGS = [
    (r"(бол|дав|жж|жм|сдав)\w*[^.]{0,20}(груд|за\s*груд)|груд\w*[^.]{0,15}(бол|дав|жж)",
     "боль или давление в груди"),
    (r"одышк|задыха|не могу дышать|нехватка воздуха|тяжело дышать|удушь",
     "одышка"),
    (r"слабость в (рук|ног)|онемел|отнял\w*сь|парал|не слушается рука",
     "слабость или онемение конечности"),
    (r"наруш\w*[^.]{0,10}реч|невнятн\w*[^.]{0,10}реч|не могу говорить|перекос\w*[^.]{0,10}лиц",
     "нарушение речи или асимметрия лица"),
    (r"обморок|потер\w*[^.]{0,10}сознан|теря\w*[^.]{0,10}сознан",
     "потеря сознания"),
    (r"кровотеч|рвота с кровью|кровь в стуле|сильно кровит|рвёт кровью",
     "кровотечение"),
    (r"судорог|припадок", "судороги"),
    (r"(?:темпер\w*|[tт])\D{0,10}(?:39|40|41)", "температура 39 и выше"),
    (r"суицид|покончить с собой|не хочу жить", "суицидальные мысли"),
    (r"ригидность затылоч|не могу наклонить голову|светобоязн", "менингеальные признаки"),
]

# Препараты и вещества, меняющие тактику. Кокаин + боль в груди — классика.
SUBSTANCES = [
    (r"кокаин|амфетамин|мефедрон|метамфетамин|соль\w*\s*(?:курит|нюха)", "стимуляторы"),
    (r"варфарин|ксарелто|прадакса|эликвис|апиксабан|ривароксабан", "антикоагулянты"),
    (r"преднизолон|метипред|дексаметазон", "системные глюкокортикоиды"),
]

HIGH_TEMP = re.compile(r"(?:темпер\w*|[tт])\D{0,10}(3[89](?:[.,]\d)?)", re.I)

CATEGORIES = {
    "emergency": {"code": "emergency", "label": "Неотложно", "rank": 0},
    "today": {"code": "today", "label": "Сегодня", "rank": 1},
    "planned": {"code": "planned", "label": "Плановый приём", "rank": 2},
    "selfcare": {"code": "selfcare", "label": "Самопомощь", "rank": 3},
}


@dataclass
class Verdict:
    urgency: dict
    red_flags: list[str]
    substances: list[str]
    reason: str
    pain: int | None
    duration_days: float | None
    duration_note: str
    duration_confident: bool
    needs_clarification: list[str]


def parse_pain(raw: str | None) -> int | None:
    if not raw:
        return None
    match = re.search(r"\d{1,2}", str(raw))
    if not match:
        return None
    return max(0, min(10, int(match.group())))


def _haystack(answers: dict) -> str:
    parts = []
    for key, value in answers.items():
        text = str(value or "").strip()
        # «нет» в ответе про тревожные признаки не должно ловиться паттернами.
        if key == "redflags" and NEGATIVE.match(text):
            continue
        parts.append(text)
    return "\n".join(parts)


def detect_red_flags(answers: dict) -> list[str]:
    hay = _haystack(answers)
    return [label for pattern, label in RED_FLAGS if re.search(pattern, hay, re.I)]


def detect_substances(answers: dict) -> list[str]:
    hay = " ".join(str(answers.get(k) or "") for k in ("meds", "complaint"))
    return [label for pattern, label in SUBSTANCES if re.search(pattern, hay, re.I)]


def classify(answers: dict, today: date | None = None) -> Verdict:
    pain = parse_pain(answers.get("pain"))
    d = dur.parse(str(answers.get("duration") or ""), today)
    red_flags = detect_red_flags(answers)
    substances = detect_substances(answers)
    hay = _haystack(answers)

    clarify: list[str] = []
    if d.days is None:
        clarify.append("длительность не разобрана — уточнить у пациента")
    elif not d.confident:
        clarify.append(f"длительность понята неточно: {d.note}")
    if pain is None:
        clarify.append("интенсивность боли не указана")

    temp = None
    match = HIGH_TEMP.search(hay)
    if match:
        temp = float(match.group(1).replace(",", "."))

    # Стимуляторы плюс боль в груди — сразу неотложно, даже если пациент
    # не назвал боль сильной: риск ишемии не коррелирует с оценкой по шкале.
    chest = any("груди" in flag for flag in red_flags)
    if substances and "стимуляторы" in substances and chest:
        red_flags.append("боль в груди на фоне стимуляторов")

    if red_flags:
        return _verdict("emergency", red_flags, substances, pain, d, clarify,
                        "Тревожные признаки: " + ", ".join(red_flags))

    if pain is not None and pain >= 7:
        reason = "Выраженная боль (%d из 10) — осмотр в течение суток." % pain
        return _verdict("today", red_flags, substances, pain, d, clarify, reason)
    if temp is not None and temp >= 38.5:
        return _verdict("today", red_flags, substances, pain, d, clarify,
                        f"Температура {temp} — осмотр в течение суток.")
    if dur.is_sudden(str(answers.get("duration") or "")):
        return _verdict("today", red_flags, substances, pain, d, clarify,
                        "Острое начало — осмотр в течение суток.")
    if substances:
        return _verdict("today", red_flags, substances, pain, d, clarify,
                        "Принимаемые препараты меняют тактику: " + ", ".join(substances))

    if (pain is not None and pain >= 3) or (d.days is not None and d.days >= 7):
        bits = []
        if pain is not None and pain >= 3:
            bits.append(f"боль {pain} из 10")
        if d.days is not None and d.days >= 7:
            bits.append(f"длительность {d.note}")
        return _verdict("planned", red_flags, substances, pain, d, clarify,
                        "Плановый приём: " + ", ".join(bits) + ".")

    if d.days is None and pain is None:
        return _verdict("planned", red_flags, substances, pain, d, clarify,
                        "Недостаточно данных для отнесения к самопомощи — плановый приём.")

    return _verdict("selfcare", red_flags, substances, pain, d, clarify,
                    "Тревожных признаков нет, интенсивность низкая — наблюдение.")


def _verdict(code, red_flags, substances, pain, d, clarify, reason) -> Verdict:
    return Verdict(
        urgency=CATEGORIES[code],
        red_flags=red_flags,
        substances=substances,
        reason=reason,
        pain=pain,
        duration_days=None if d.days is None else round(d.days, 3),
        duration_note=d.note,
        duration_confident=d.confident,
        needs_clarification=clarify,
    )


def to_dict(verdict: Verdict) -> dict:
    return asdict(verdict)
